client/keys/new.go

```diff
@@ -0,0 +1,186 @@
+			return err
+		}
+	}
+
+	flags := cmd.Flags()
+	useDefaults, _ := flags.GetBool(flagNewDefault)
+	bipFlag := flags.Lookup(flagBIP44Path)
+
+	bip44Params, err := getBIP44ParamsAndPath(bipFlag.Value.String(), bipFlag.Changed || useDefaults)
+	if err != nil {
+		return err
+	}
+
+	// if we're using ledger, only thing we need is the path.
+	// generate key and we're done.
+	if viper.GetBool(client.FlagUseLedger) {
+
+		algo := keys.Secp256k1               // SigningAlgo(viper.GetString(flagType))
+		path := bip44Params.DerivationPath() // ccrypto.DerivationPath{44, 118, account, 0, index}
+		info, err := kb.CreateLedger(name, path, algo)
+		if err != nil {
+			return err
+		}
+		printCreate(info, "")
+		return nil
+	}
+
+	// get the mnemonic
+	var mnemonic string
+	if !useDefaults {
+		mnemonic, err = client.GetString("> Enter your bip39 mnemonic, or hit enter to generate one.", buf)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(mnemonic) == 0 {
+		// read entropy seed straight from crypto.Rand and convert to mnemonic
+		entropySeed, err := bip39.NewEntropy(mnemonicEntropySize)
+		if err != nil {
+			return err
+		}
+		mnemonic, err = bip39.NewMnemonic(entropySeed[:])
+		if err != nil {
+			return err
+		}
+	}
+
+	// get bip39 passphrase
+	var bip39Passphrase string
+	if !useDefaults {
+		printStep()
+		printPrefixed("Enter your bip39 passphrase. This is combined with the mnemonic to derive the seed")
+		bip39Passphrase, err = client.GetString("> Most users should just hit enter to use the default, \"\"", buf)
+		if err != nil {
+			return err
+		}
+
+		// if they use one, make them re-enter it
+		if len(bip39Passphrase) != 0 {
+			p2, err := client.GetString("Repeat the passphrase:", buf)
+			if err != nil {
+				return err
+			}
+			if bip39Passphrase != p2 {
+				return errors.New("passphrases don't match")
+			}
+		}
+	}
+
+	// get the encryption password
+	printStep()
+	encryptPassword, err := client.GetCheckPassword(
+		"> Enter a passphrase to encrypt your key to disk:",
+		"> Repeat the passphrase:", buf)
+	if err != nil {
+		return err
+	}
+
+	info, err := kb.Derive(name, mnemonic, bip39Passphrase, encryptPassword, *bip44Params)
+	if err != nil {
+		return err
+	}
+	_ = info
+
+	return nil
+}
+
+func getBIP44ParamsAndPath(path string, flagSet bool) (*hd.BIP44Params, error) {
+	buf := bufio.NewReader(os.Stdin)
+	bip44Path := path
+
+	// if it wasnt set in the flag, give it a chance to overide interactively
+	if !flagSet {
+		printStep()
+
+		var err error
+		bip44Path, err = client.GetString(fmt.Sprintf("> Enter your bip44 path. Default is %s\n", path), buf)
+		if err != nil {
+			return nil, err
+		}
+		if len(bip44Path) == 0 {
+			bip44Path = path
+		}
+	}
+
+	bip44params, err := hd.NewParamsFromPath(bip44Path)
+	if err != nil {
+		return nil, err
+	}
+	return bip44params, nil
+}
+
+func printPrefixed(msg string) {
+	fmt.Printf("> %s\n", msg)
+}
+
+func printStep() {
+	printPrefixed("-------------------------------------")
+}
```

Comment:
funny print functions - I'd almost like to see them someplace common outside of this package eventually. Fine for now though